cpp/src/arrow/builder.cc

```diff
@@ -721,11 +671,19 @@ Status BooleanBuilder::Resize(int64_t capacity) {
     capacity = kMinBuilderCapacity;
   }
 
+  const int64_t nbytes = BitUtil::BytesForBits(capacity);
   if (capacity_ == 0) {
-    RETURN_NOT_OK(Init(capacity));
+    RETURN_NOT_OK(AllocateResizableBuffer(pool_, nbytes, &data_));
+    raw_data_ = reinterpret_cast<uint8_t*>(data_->mutable_data());
+
+    // We zero the memory for booleans to keep things simple; for some reason if
+    // we do not, even though we may write every bit (through in-place | or &),
+    // valgrind will still show a warning. If we do not zero the bytes here, we
+    // will have to be careful to zero them in AppendNull and AppendNulls. Also,
+    // zeroing the bits results in deterministic bits when each byte may have a
+    // mix of nulls and not nulls.
+    memset(raw_data_, 0, static_cast<size_t>(nbytes));
```

Comment:
The same situation as in ArrayBuilder::Resize. Here we zero up-to size in the "init phase", instead of capacity, like in ArrayBuilder::Resize. Later we zero between the old size and new size. I'd say it doesn't matter that much if we zero padding twice, as long as it's done the same way everywhere.

Reply:
Done. I am doing a little bit of cleaning and renaming of variables for clarity also, since all these different meanings for "capacity" and "size" are sure to cause programmer error otherwise